### PR TITLE
[core] support attention backends for LTX

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Working code:

<details>
<summary>Unfold</summary>

```py
import torch
from diffusers import LTXConditionPipeline
from diffusers.pipelines.ltx.pipeline_ltx_condition import LTXVideoCondition
from diffusers.utils import export_to_video, load_image, load_video

def round_to_nearest_resolution_acceptable_by_vae(height, width):
    height = height - (height % pipe.vae_spatial_compression_ratio)
    width = width - (width % pipe.vae_spatial_compression_ratio)
    return height, width

pipe = LTXConditionPipeline.from_pretrained(
    "Lightricks/LTX-Video-0.9.8-13B-distilled", torch_dtype=torch.bfloat16
).to("cuda")

image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/penguin.png")
video = load_video(export_to_video([image])) # compress the image using video compression as the model was trained on videos
condition1 = LTXVideoCondition(video=video, frame_index=0)
print(f"{image.size=}")

prompt = "A cute little penguin takes out a book and starts reading it"
negative_prompt = "worst quality, inconsistent motion, blurry, jittery, distorted"
expected_height, expected_width = 480, 832
downscale_factor = 2 / 3
num_frames = 96

# Part 1. Generate video at smaller resolution
downscaled_height, downscaled_width = int(expected_height * downscale_factor), int(expected_width * downscale_factor)
downscaled_height, downscaled_width = round_to_nearest_resolution_acceptable_by_vae(
    downscaled_height, downscaled_width
)
video = pipe(
    conditions=[condition1],
    prompt=prompt,
    negative_prompt=negative_prompt,
    width=expected_width,
    height=expected_height,
    num_frames=num_frames,
    num_inference_steps=8,
    generator=torch.Generator().manual_seed(0),
).frames[0]
export_to_video(video, "output.mp4", fps=24)
```

</details>

<details>
<summary>Output</summary>

https://github.com/user-attachments/assets/219086ba-08cd-4856-8e3d-5be7f005a320

</details>

Regarding the output:

1. I get the same output on `main`.
2. I didn't follow the three-stage pipeline of first generating low-res latents, upsampling them, and then running a few rounds of denoising. So, the quality loss is somewhat expected I believe (regardless of this PR).

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_ltx.py [hunk 1]
```diff
@@ -37,20 +37,30 @@
 
 
 class LTXVideoAttentionProcessor2_0:
+    def __new__(cls, *args, **kwargs):
+        deprecation_message = "`LTXVideoAttentionProcessor2_0` is deprecated and this will be removed in a future version. Please use `LTXVideoAttentionProcessor2_0`"
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/transformer_ltx.py]:
```suggestion
        deprecation_message = "`LTXVideoAttentionProcessor2_0` is deprecated and this will be removed in a future version. Please use `LTXVideoAttnProcessor`"
```

File: src/diffusers/models/transformers/transformer_ltx.py [hunk 2]
```diff
@@ -37,20 +37,30 @@
...
+        deprecate("LTXVideoAttentionProcessor2_0", "1.0.0", deprecation_message)
+
+        return LTXAttnProcessor(*args, **kwargs)
+
+
+class LTXAttnProcessor:
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/transformer_ltx.py]:
```suggestion
class LTXVideoAttnProcessor:
```

File: src/diffusers/models/transformers/transformer_ltx.py [hunk 3]
```diff
@@ -78,21 +88,90 @@ def __call__(
...
+        dropout: float = 0.0,
+        bias: bool = True,
+        cross_attention_dim: Optional[int] = None,
+        out_bias: bool = True,
+        qk_norm: str = "rms_norm_across_heads",
+        processor=None,
+    ):
+        super().__init__()
+        if qk_norm != "rms_norm_across_heads":
+            raise NotImplementedError
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/transformer_ltx.py]: Maybe elaborate here with a message saying no other qk norm is supported in our implementation

File: src/diffusers/models/transformers/transformer_ltx.py [hunk 4]
```diff
@@ -78,21 +88,90 @@ def __call__(
...
+        self.inner_dim = dim_head * heads
+        self.inner_kv_dim = self.inner_dim if kv_heads is None else dim_head * kv_heads
+        self.query_dim = query_dim
+        self.cross_attention_dim = cross_attention_dim if cross_attention_dim is not None else query_dim
+        self.use_bias = bias
+        self.dropout = dropout
+        self.out_dim = query_dim
+        self.heads = heads
+
+        norm_eps, norm_elementwise_affine = 1e-5, True
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/transformer_ltx.py]: (nit): let's do one of these in a line

File: src/diffusers/models/transformers/transformer_ltx.py [hunk 5]
```diff
@@ -78,21 +88,90 @@ def __call__(
...
-        hidden_states = hidden_states.transpose(1, 2).flatten(2, 3)
+        hidden_states = hidden_states.flatten(2, 3)
         hidden_states = hidden_states.to(query.dtype)
 
         hidden_states = attn.to_out[0](hidden_states)
         hidden_states = attn.to_out[1](hidden_states)
         return hidden_states
 
 
+class LTXAttention(torch.nn.Module, AttentionModuleMixin):
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/transformer_ltx.py:112]: super clean!

File: src/diffusers/models/transformers/transformer_ltx.py [hunk 6]
```diff
@@ -78,21 +88,90 @@ def __call__(
...
+        image_rotary_emb: Optional[torch.Tensor] = None,
+        **kwargs,
+    ) -> torch.Tensor:
+        attn_parameters = set(inspect.signature(self.processor.__call__).parameters.keys())
+        unused_kwargs = [k for k, _ in kwargs.items() if k not in attn_parameters]
+        if len(unused_kwargs) > 0:
+            logger.warning(
+                f"joint_attention_kwargs {unused_kwargs} are not expected by {self.processor.__class__.__name__} and will be ignored."
+            )
+        kwargs = {k: w for k, w in kwargs.items() if k in attn_parameters}
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/transformer_ltx.py]: In LTX case, I think we can simply remove these lines, since we never supported passing attention kwargs in the first place. These would result in a no-op I believe

**Comment by sayakpaul** [comment]
Thanks @a-r-r-o-w! Have you started Wan already? If so, cool! If not, can I start and test your parallelism PR a bit with that?

**Comment by a-r-r-o-w** [comment]
Wan attention backend support has already been merged in #11918. For testing parallelism, I've tested most of the implementations (even some outside diffusers) to validate the soundness of going forth with CP-plans. Some plans are available here: https://github.com/huggingface/finetrainers/blob/f476c3717da6cbfb1070505a99ee989426b46d9c/finetrainers/models/_metadata/transformer.py#L68

**Comment by sayakpaul** [comment]
Cool! Once this gets matured a bit would love documenting some benchmakrs across the board.
